File: app/src/main/java/com/cajetan/youtubeplus/FavouriteListAdapter.java
package com.cajetan.youtubeplus;

import android.support.v7.widget.RecyclerView;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Color;
import android.support.annotation.NonNull;
import android.util.DisplayMetrics;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.google.api.services.youtube.model.Video;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.List;

// TODO: change ViewHolder colour onClick

public class FavouriteListAdapter extends RecyclerView.Adapter<FavouriteListAdapter.VideoViewHolder> {
    private static final String TAG = FavouriteListAdapter.class.getSimpleName();

    private final ListItemClickListener mOnClickListener;
    private ArrayList<Video> mVideos;
    private Context context;

    /*//////////////////////////////////////////////////////////////////////////////
    // Lifecycle
    //////////////////////////////////////////////////////////////////////////////*/

    FavouriteListAdapter(List<Video> videos, ListItemClickListener listener, Context context) {
        mOnClickListener = listener;
        mVideos = new ArrayList<>(videos);
        this.context = context;
    }

    @Override
    @NonNull
    public VideoViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View view = inflater.inflate(R.layout.video_list_item, parent, false);
        return new VideoViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull VideoViewHolder holder, int position) {
        holder.bind(mVideos.get(position));
    }

    /*//////////////////////////////////////////////////////////////////////////////
    // Utils
    //////////////////////////////////////////////////////////////////////////////*/

    @Override
    public int getItemCount() {
        return mVideos.size();
    }

    public void addItems(List<Video> items) {
        this.mVideos.addAll(items);
        notifyDataSetChanged();
    }

    public void clearItems() {
        this.mVideos = new ArrayList<>();
        notifyDataSetChanged();
    }

    private static int dpToPixel(float dp, Context context) {
        Resources resources = context.getResources();
        DisplayMetrics metrics = resources.getDisplayMetrics();
        return Math.round(dp * ((float) metrics.densityDpi / DisplayMetrics.DENSITY_DEFAULT));
    }

    /*//////////////////////////////////////////////////////////////////////////////
    // Callbacks & others
    //////////////////////////////////////////////////////////////////////////////*/

    class VideoViewHolder extends RecyclerView.ViewHolder
            implements View.OnClickListener, View.OnLongClickListener {

        TextView videoTitleView;
        TextView videoChannelView;
        TextView videoDurationView;
        ImageView videoThumbnailView;

        VideoViewHolder(View itemView) {
            super(itemView);

            videoTitleView = itemView.findViewById(R.id.video_title);
            videoChannelView = itemView.findViewById(R.id.video_author);
            videoThumbnailView = itemView.findViewById(R.id.video_thumbnail);
            videoDurationView = itemView.findViewById(R.id.video_duration);

            itemView.setOnClickListener(this);
            itemView.setOnLongClickListener(this);
        }

        void bind(Video video) {
            videoTitleView.setText(video.getSnippet().getTitle());
            videoChannelView.setText(video.getSnippet().getChannelTitle());
            videoThumbnailView.setBackgroundColor(Color.parseColor("#e5e5e5"));

            String duration = video.getContentDetails().getDuration();

            // Show duration in HH:MM:SS or 'LIVE' accordingly
            if (duration.equals("PT0S")) {
                videoDurationView.setBackgroundColor(Color.RED);
                videoDurationView.setTextColor(Color.WHITE);
                videoDurationView.setText("LIVE");
            } else {
                duration = duration.substring(2, duration.length() - 1);
                duration = duration.replace("M", ":");
                videoDurationView.setText(duration);
            }

            String thumbnailUrl = null;

            if (video.getSnippet().getThumbnails().getMaxres() != null)
                thumbnailUrl = video.getSnippet().getThumbnails().getMaxres().getUrl();
            else if (video.getSnippet().getThumbnails().getHigh() != null)
                thumbnailUrl = video.getSnippet().getThumbnails().getHigh().getUrl();
            else if (video.getSnippet().getThumbnails().getMedium() != null)
                thumbnailUrl = video.getSnippet().getThumbnails().getMedium().getUrl();
            else if (video.getSnippet().getThumbnails().getStandard() != null)
                thumbnailUrl = video.getSnippet().getThumbnails().getStandard().getUrl();
            else if (video.getSnippet().getThumbnails().getDefault() != null)
                thumbnailUrl = video.getSnippet().getThumbnails().getDefault().getUrl();

            if (thumbnailUrl != null) {
                Picasso.get().load(thumbnailUrl)
                        .resize(dpToPixel(160, context), dpToPixel(90, context))
                        .centerCrop().into(videoThumbnailView);
            }
        }

        @Override
        public void onClick(View v) {
            mOnClickListener.onListItemClick(mVideos.get(getAdapterPosition()).getId());
        }

        @Override
        public boolean onLongClick(View v) {
            mOnClickListener.onListItemLongClick(mVideos.get(getAdapterPosition()).getId());
            return true;
        }
    }

    public interface ListItemClickListener {
        void onListItemClick(String clickedVideoId);
        void onListItemLongClick(String clickedVideoId);
    }
}

File: app/src/main/java/com/cajetan/youtubeplus/FavouriteListAdapter.kt
package com.cajetan.youtubeplus

import android.content.Context
import android.graphics.Color
import android.os.Build
import androidx.recyclerview.widget.RecyclerView
import android.util.DisplayMetrics
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.google.api.services.youtube.model.Video
import com.squareup.picasso.Picasso

class FavouriteListAdapter(videos: List<Video>, listener: ListItemClickListener, context: Context) :
        RecyclerView.Adapter<FavouriteListAdapter.VideoViewHolder>() {

    private val mVideos: ArrayList<Video> = ArrayList(videos)
    private val mOnClickListener = listener
    private val mContext = context

    ////////////////////////////////////////////////////////////////////////////////
    // Lifecycle
    ////////////////////////////////////////////////////////////////////////////////

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VideoViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val view = inflater.inflate(R.layout.video_list_item, parent, false)
        return VideoViewHolder(view)
    }

    override fun onBindViewHolder(holder: VideoViewHolder, position: Int) {
        holder.bind(mVideos[position])
    }

    ////////////////////////////////////////////////////////////////////////////////
    // Utils
    ////////////////////////////////////////////////////////////////////////////////

    override fun getItemCount(): Int {
        return mVideos.size
    }

    fun addItems(items: List<Video>) {
        mVideos.addAll(items)
        notifyDataSetChanged()
    }

    fun clearItems() {
        mVideos.clear()
        notifyDataSetChanged()
    }

    fun filterItems(query: String) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N)
            mVideos.removeIf { !it.snippet.title.toLowerCase().contains(query.toLowerCase()) }
        notifyDataSetChanged()

//        TODO: implement functionality for older versions
    }

    private fun dpToPixel(dp: Float, context: Context): Int {
        val metrics = context.resources.displayMetrics
        return Math.round(dp * (metrics.densityDpi / DisplayMetrics.DENSITY_DEFAULT))
    }

    private fun parseDuration(input: String): String {
        var result = input.filterNot { it == 'P' || it == 'T' }

        var hours: Int = when {
            result.contains('D') -> {
                val temp = result.takeWhile { it != 'D' }.toInt() * 24
                result = result.dropWhile { it != 'D' }.filterNot { it == 'D' }
                temp
            }
            else -> 0
        }

        hours += when {
            result.contains('H') -> {
                val temp = result.takeWhile { it != 'H' }.toInt()
                result = result.dropWhile { it != 'H' }.filterNot { it == 'H' }
                temp
            }
            else -> 0
        }

        val minutes: Int = when {
            result.contains('M') -> {
                val temp = result.takeWhile { it != 'M' }.toInt()
                result = result.dropWhile { it != 'M' }.filterNot { it == 'M' }
                temp
            }
            else -> 0
        }

        val seconds: Int = when {
            result.contains('S') -> result.takeWhile { it != 'S' }.toInt()
            else -> 0
        }

        return when (hours) {
            0 -> "$minutes:${String.format("%02d", seconds)}"
            else -> "$hours:${String.format("%02d", minutes)}:${String.format("%02d", seconds)}"
        }
    }

    ////////////////////////////////////////////////////////////////////////////////
    // Callbacks & others
    ////////////////////////////////////////////////////////////////////////////////

    inner class VideoViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView),
            View.OnClickListener, View.OnLongClickListener {

        private val videoTitleView: TextView = itemView.findViewById(R.id.video_title)
        private val videoChannelView: TextView = itemView.findViewById(R.id.video_author)
        private val videoDurationView: TextView = itemView.findViewById(R.id.video_duration)
        private val videoThumbnailView: ImageView = itemView.findViewById(R.id.video_thumbnail)

        init {
            itemView.setOnClickListener(this)
            itemView.setOnLongClickListener(this)
        }

        fun bind(video: Video) {
            videoTitleView.text = video.snippet.title
            videoChannelView.text = video.snippet.channelTitle
            videoThumbnailView.setBackgroundResource(R.color.darkerLightGrey)

            val duration = video.contentDetails.duration
            videoDurationView.text = when (duration) {
                "PT0S" -> {
                    videoDurationView.setBackgroundColor(Color.RED)
                    videoDurationView.setTextColor(Color.WHITE)
                    mContext.getString(R.string.live)
                }
                else -> parseDuration(duration)
            }

            val thumbnailUrl: String = when {
                video.snippet.thumbnails.maxres != null -> video.snippet.thumbnails.maxres.url
                video.snippet.thumbnails.high != null -> video.snippet.thumbnails.high.url
                video.snippet.thumbnails.medium != null -> video.snippet.thumbnails.medium.url
                video.snippet.thumbnails.standard != null -> video.snippet.thumbnails.standard.url
                video.snippet.thumbnails.default != null -> video.snippet.thumbnails.default.url
                else -> ""
            }

            if (thumbnailUrl != "") {
                Picasso.get().load(thumbnailUrl)
                        .resize(dpToPixel(160f, mContext), dpToPixel(90f, mContext))
                        .centerCrop().into(videoThumbnailView)
            }
        }

        override fun onClick(v: View?) {
            mOnClickListener.onListItemClick(mVideos[adapterPosition].id)
        }

        override fun onLongClick(v: View?): Boolean {
            mOnClickListener.onListItemLongClick(mVideos[adapterPosition].id)
            return true
        }
    }

    interface ListItemClickListener {
        fun onListItemClick(clickedVideoId: String)
        fun onListItemLongClick(clickedVideoId: String)
    }
}
